generator-engine/src/main/java/hu/blackbelt/judo/meta/psm/generator/engine/PsmGenerator.java

```diff
@@ -181,46 +181,56 @@ private static GeneratorResult<ActorType> execute(PsmModel psmModel, GeneratorPa
-                            return generatedFile;
-                        }));
+                    if (processingList == null) {
+                        log.warn("Factory exxpression is defined, but returns null - " + generatorTemplate.toString());
+                    } else {
+                        for (Object element : processingList) {
+                            tasks.add(CompletableFuture.supplyAsync(() -> {
+                                StandardEvaluationContext templateContext = defaultSpringELContextProvider.apply(element);
+                                templateContext.setVariable(ACTOR_TYPE, actorType);
+
+                                Context.Builder contextBuilder = defaultHandlebarsContextBuilder.apply(element)
+                                        .combine(ACTOR_TYPE, actorType);
+
+                                callBindContextForTypeIfCan(parameter.getGeneratorContext(), StandardEvaluationContext.class, templateContext);
+                                callBindContextForTypeIfCan(parameter.getGeneratorContext(), Map.class, parameter.getExtraContextVariables().get());
+
+                                generatorTemplate.evalToContextBuilder(templateEvaulator, contextBuilder, templateContext);
+                                GeneratedFile generatedFile = generateFile(parameter.getGeneratorContext(), templateContext, templateEvaulator, generatorTemplate, contextBuilder, log);
+                                result.getGeneratedByDiscriminator().get(actorType).add(generatedFile);
+                                return generatedFile;
+                            }));
+                        }
                     }
                 });
             } else {
                 evaulationContext.setVariable(TEMPLATE, generatorTemplate);
-                Set<?> iterableCollection = new HashSet<>(List.of(generatorTemplate));
-
+                Collection<?> processingList = new HashSet<>(Collections.singletonList(generatorTemplate));
                 if (templateEvaulator.getTemplate() != null) {
-                    iterableCollection = new HashSet<>(Collections.singletonList(model));
+                    processingList = new HashSet<>(Collections.singletonList(model));
+                }
+                if (templateEvaulator.getFactoryExpression() != null) {
+                    processingList = templateEvaulator.getFactoryExpressionResultOrValue(generatorTemplate, model, Collection.class);
                 }
+                if (processingList == null) {
+                    log.warn("Factory exxpression is defined, but returns null - " + generatorTemplate.toString());
+                } else {
 
-                for (Object element : templateEvaulator.getFactoryExpressionResultOrValue(generatorTemplate, iterableCollection, Collection.class)) {
-                    tasks.add(CompletableFuture.supplyAsync(() -> {
+                    for (Object element : processingList) {
```

Comment:
This part seems similar (might be wrong, not sure) to the changed code above, can't it be extracted as a method to prevent boiler plate?

Reply:
More bolertplate is generated for the generalization itself.